Add recipe for hatty

ErikPrantare opened the pull request:
### Brief summary of what the package does

This package renders hats over the words in buffers.  These hats may be used to index the positions.  Designed primarily for use in conjunction with voice control, allowing a token to be referred to by e.g. "yellow bolt near" (for the token with a yellow bolt over the n).  Inspired by cursorless for vscode.

### Direct link to the package repository

https://github.com/ErikPrantare/hatty.el

### Your association with the package

Author

### Relevant communications with the upstream package maintainer

None needed

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  I see this is inspired by the "cursorless" project (which is new to me) but is implementing only one component of that project's overall scheme.  Without the other components implemented (voice recognition, text manipulation?) I am wondering about whether it is time just yet to index this into MELPA.  Let me know if I have everything correct here, though, so I'm not making any undue assumptions.

In the meantime I'll also note that you don't have to specify `:branch "master"` in your recipe as long as this is your default branch.  This enables you to rename that branch in the future if you ever chose to.  It's also worth noting that including your `attribution` directory is a bit unusual, and I'd see if you can't just include that attribution in the source file itself as needed.

ErikPrantare commented:
Thank you for the feedback!

My aim with this library is mainly to provide a mechanism to construct interfaces able to overcome a fundamental issue with packages like avy.  The issue with avy is this that be mental process for acting at on a token is as follows

1. Think the action you want to perform ("I want to go to this token")
2. Perform a command (avy-goto-char-2), *wait for the computer to process your command*.
3. Parse the interface that became visible with the command ("Which keys will I have to press to go to my token?")
4. Enter index of the location

The issue lies in (2), which forces a "stop gap" between thinking the action and parsing the interface.  This already adds mental overhead with keyboard input, but even more so with voice input where the command disables you from chaining together actions in a single utterance, having instead to wait for (2) to be able to perform the action.  By rendering hats, it is possible to overcome this limitation as the indexing is always already visible.

How precisely one chooses to use this library I have left open.  I use this library as the basis for my own voice control configuration, but in the future I am e.g. looking at working with the cursorless team to wire this up with their engine and only use this as a renderer.  I also want to leave the actual input method open, allowing the design of keyboard interfaces on top of this.  This was designed to make voice control more fluid, but it is not my intention to *restrict it* to voice control.

Now, I also see the utility in providing some basic configuration that makes it simpler to use in some basic manner out of the box.  I am not sure what a good approach for this would be, and am open for any suggestions.  In any case, it would probably be a good idea to clarify the scope of the library in its README: It is not meant to be a cursorless, but to enable the construction of cursorless-like interfaces.

---

Thank you for the heads-up about specifying the master branch, I will fix that.  As for the attribution, I agree with you that it is awkward.  I also think having it as a comment directly in the source is a bit awkward, but you are probably right that it is a better approach.  I will take care of that as well.

ErikPrantare commented:
I have done the following:
- Moved attribution to the source file.
- Removed the attribution directory from the recipe.
- Added introductory documentation stating the purpose of the package.
- Removed specification of branch in recipe.

ErikPrantare commented:
Update: I am currently working on building a package for cursorless-style editing facilities upon this.  Would you recommend keeping the separation between the two packages, or merge them into one?

riscy commented:
> My aim with this library is mainly to provide a mechanism to construct interfaces

This is definitely commendable and I'm excited to see where this goes!

> I am currently working on building a package for cursorless-style editing facilities upon this. Would you recommend keeping the separation between the two packages, or merge them into one?

Here's my thinking on this...  I think it's sometimes worth delaying indexing a pure dependency like this into the MELPA index without anything built on top of it.  Of course it's also partly a question of how specialized the dependency is - packages like s- or dash or f- are very general, but I'd call yours very specialized.

The reason it's worth delaying is because you probably *don't* want other people to start depending on this stuff while you're figuring out your own application, which is ultimately what this stuff is built for.  You want to be able to change the interfaces, you want to be able to make breaking changes, you may decide certain features don't belong anymore (or belong in your application instead).  Or life may intervene, as it does, and you may never complete the application.

This was so long-winded, but I think it's just fine to maintain the separation between the two packages - but let's wait until there are actually two packages to merge.

Let me know your thoughts on that -

ErikPrantare commented:
All right, should I get back here when I have a working prototype of a system using this as a dependency?

riscy commented:
Ideally when there are two packages you want to merge, rather than just a prototype of a system.

riscy commented:
I'll mark this awaiting-upstream for now, but feel free to reach out.

ErikPrantare commented:
Update: The system I use daily now that builds upon this resides here: https://github.com/ErikPrantare/cursorfree.el

It works for my usage, but I am looking to make sure that it works well for at least one other person too.  Just thought I would update that this is still progressing.

riscy commented:
Thank you for the update, @ErikPrantare!  Let's get this hatty dependency merged.  Here's all that's left:

⸺ `hatty.el` with byte-compile using Emacs 29.4:
```
hatty.el:51:20: Warning: the function `modus-themes-color' is not known to be defined.
```
I think especially if you depend only on Emacs 26.1, you should declare a dependency on modus-themes and `(require ..)` it to resolve this warning.

⸺ `hatty.el` with checkdoc 0.6.2 (fix *within reason*):
```
hatty.el:198: Documentation strings should not start or end with whitespace
```

⸺ Package and license:
- In case you haven't, ensure GitHub release v0.3.1 is up-to-date with your current code and `Package-Version`

riscy commented:
Looks great.  Thank you for your patience with all of this - and welcome to MELPA!
